File: apps/desktop/src-tauri/src/db/schema.rs
use rusqlite::{Connection, Result};

pub fn run_migrations(conn: &Connection) -> Result<()> {
    conn.execute_batch("PRAGMA journal_mode=WAL;")?;
    conn.execute_batch("PRAGMA foreign_keys=ON;")?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS dictionaries (
            id TEXT PRIMARY KEY NOT NULL,
            name TEXT NOT NULL UNIQUE,
            language_from TEXT NOT NULL,
            language_to TEXT NOT NULL,
            source_filename TEXT NOT NULL,
            source_path TEXT,
            import_status TEXT NOT NULL DEFAULT 'queued'
                CHECK(import_status IN ('queued', 'importing', 'ready', 'failed')),
            entry_count INTEGER NOT NULL DEFAULT 0,
            last_error TEXT,
            created_at TEXT NOT NULL,
            updated_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS dictionary_entries (
            id TEXT PRIMARY KEY NOT NULL,
            dictionary_id TEXT NOT NULL REFERENCES dictionaries(id) ON DELETE CASCADE,
            headword TEXT NOT NULL,
            normalized_headword TEXT NOT NULL,
            transcription TEXT,
            definition_text TEXT NOT NULL,
            example_text TEXT,
            audio_reference TEXT,
            source_order INTEGER NOT NULL DEFAULT 0,
            created_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_dictionary_entries_dictionary_id
            ON dictionary_entries(dictionary_id);",
    )?;

    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_dictionary_entries_normalized_headword
            ON dictionary_entries(normalized_headword);",
    )?;

    conn.execute_batch(
        "CREATE VIRTUAL TABLE IF NOT EXISTS dictionary_entries_fts
            USING fts5(
                normalized_headword,
                content=dictionary_entries,
                content_rowid=rowid
            );",
    )?;

    // FTS triggers for keeping the index in sync
    conn.execute_batch(
        "CREATE TRIGGER IF NOT EXISTS dictionary_entries_fts_insert
            AFTER INSERT ON dictionary_entries BEGIN
                INSERT INTO dictionary_entries_fts(rowid, normalized_headword)
                VALUES (new.rowid, new.normalized_headword);
            END;

        CREATE TRIGGER IF NOT EXISTS dictionary_entries_fts_delete
            AFTER DELETE ON dictionary_entries BEGIN
                INSERT INTO dictionary_entries_fts(dictionary_entries_fts, rowid, normalized_headword)
                VALUES ('delete', old.rowid, old.normalized_headword);
            END;

        CREATE TRIGGER IF NOT EXISTS dictionary_entries_fts_update
            AFTER UPDATE ON dictionary_entries BEGIN
                INSERT INTO dictionary_entries_fts(dictionary_entries_fts, rowid, normalized_headword)
                VALUES ('delete', old.rowid, old.normalized_headword);
                INSERT INTO dictionary_entries_fts(rowid, normalized_headword)
                VALUES (new.rowid, new.normalized_headword);
            END;",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS cards (
            id TEXT PRIMARY KEY NOT NULL,
            language TEXT NOT NULL,
            headword TEXT NOT NULL,
            answer_text TEXT NOT NULL,
            example_text TEXT,
            notes TEXT,
            audio_reference TEXT,
            source_entry_ids TEXT NOT NULL DEFAULT '[]',
            learning_status TEXT NOT NULL DEFAULT 'unreviewed'
                CHECK(learning_status IN ('unreviewed', 'learned', 'not_learned')),
            created_at TEXT NOT NULL,
            updated_at TEXT NOT NULL,
            last_reviewed_at TEXT,
            UNIQUE(headword, language)
        );",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS collections (
            id TEXT PRIMARY KEY NOT NULL,
            name TEXT NOT NULL UNIQUE,
            description TEXT,
            created_at TEXT NOT NULL,
            updated_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS collection_memberships (
            collection_id TEXT NOT NULL REFERENCES collections(id) ON DELETE CASCADE,
            card_id TEXT NOT NULL REFERENCES cards(id) ON DELETE CASCADE,
            created_at TEXT NOT NULL,
            PRIMARY KEY (collection_id, card_id)
        );",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS review_events (
            id TEXT PRIMARY KEY NOT NULL,
            card_id TEXT NOT NULL REFERENCES cards(id) ON DELETE CASCADE,
            result TEXT NOT NULL CHECK(result IN ('learned', 'not_learned')),
            reviewed_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS languages (
            code TEXT PRIMARY KEY NOT NULL
                CHECK(LENGTH(code) = 2 AND code = LOWER(code)),
            title TEXT NOT NULL,
            created_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "INSERT OR IGNORE INTO languages (code, title, created_at)
         VALUES ('en', 'English', datetime('now'));",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS ai_credentials (
            id TEXT PRIMARY KEY,
            provider TEXT NOT NULL,
            label TEXT,
            is_active INTEGER NOT NULL DEFAULT 1 CHECK (is_active IN (0, 1)),
            created_at TEXT NOT NULL,
            updated_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "CREATE UNIQUE INDEX IF NOT EXISTS ux_ai_credentials_provider_active
         ON ai_credentials(provider)
         WHERE is_active = 1;",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS chat_conversations (
            id TEXT PRIMARY KEY NOT NULL,
            title TEXT NOT NULL,
            model TEXT,
            collection_id TEXT,
            created_at TEXT NOT NULL,
            updated_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS chat_messages (
            id TEXT PRIMARY KEY NOT NULL,
            conversation_id TEXT NOT NULL REFERENCES chat_conversations(id) ON DELETE CASCADE,
            role TEXT NOT NULL CHECK(role IN ('user', 'assistant')),
            body TEXT NOT NULL,
            metadata_json TEXT,
            created_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_chat_messages_conversation_created
            ON chat_messages(conversation_id, created_at);",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS custom_chat_models (
            id TEXT PRIMARY KEY,
            name TEXT NOT NULL UNIQUE,
            title TEXT NOT NULL,
            provider TEXT NOT NULL,
            created_at TEXT NOT NULL
        );",
    )?;

    // -----------------------------------------------------------------------
    // Sync metadata tables
    // -----------------------------------------------------------------------

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS sync_devices (
            id TEXT PRIMARY KEY NOT NULL,
            display_name TEXT NOT NULL,
            host TEXT,
            port INTEGER,
            is_local INTEGER NOT NULL DEFAULT 0 CHECK(is_local IN (0,1)),
            credential TEXT,
            last_sync_request_id TEXT,
            paired_at TEXT,
            last_sync_at TEXT,
            created_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS sync_changes (
            id TEXT PRIMARY KEY NOT NULL,
            device_id TEXT NOT NULL,
            table_name TEXT NOT NULL,
            row_id TEXT NOT NULL,
            op_type TEXT NOT NULL CHECK(op_type IN ('insert','update')),
            logical_clock INTEGER NOT NULL,
            payload_json TEXT NOT NULL,
            request_id TEXT,
            applied_at TEXT,
            created_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_sync_changes_device_clock
            ON sync_changes(device_id, logical_clock);",
    )?;
    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_sync_changes_table_row
            ON sync_changes(table_name, row_id);",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS sync_tombstones (
            id TEXT PRIMARY KEY NOT NULL,
            device_id TEXT NOT NULL,
            table_name TEXT NOT NULL,
            row_id TEXT NOT NULL,
            logical_clock INTEGER NOT NULL,
            request_id TEXT,
            deleted_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_sync_tombstones_device_clock
            ON sync_tombstones(device_id, logical_clock);",
    )?;
    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_sync_tombstones_table_row
            ON sync_tombstones(table_name, row_id);",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS sync_cursors (
            peer_device_id TEXT PRIMARY KEY NOT NULL,
            last_remote_logical_clock INTEGER NOT NULL DEFAULT 0,
            last_acknowledged_local_logical_clock INTEGER NOT NULL DEFAULT 0,
            updated_at TEXT NOT NULL
        );",
    )?;

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS sync_state (
            id TEXT PRIMARY KEY NOT NULL DEFAULT 'local',
            first_successful_sync_at TEXT
        );",
    )?;

    conn.execute(
        "INSERT OR IGNORE INTO sync_state (id, first_successful_sync_at) VALUES ('local', NULL)",
        [],
    )?;

    // Desktop-side sync triggers (same tables as mobile, mirrors mobile trigger logic)
    conn.execute_batch(
        "CREATE TRIGGER IF NOT EXISTS sync_trig_insert_cards
         AFTER INSERT ON cards
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_changes (id, device_id, table_name, row_id, op_type, logical_clock, payload_json, created_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'cards', NEW.id, 'insert',
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             json_object('id',NEW.id,'language',NEW.language,'headword',NEW.headword,'answer_text',NEW.answer_text,'example_text',NEW.example_text,'notes',NEW.notes,'audio_reference',NEW.audio_reference,'source_entry_ids',NEW.source_entry_ids,'learning_status',NEW.learning_status,'created_at',NEW.created_at,'updated_at',NEW.updated_at,'last_reviewed_at',NEW.last_reviewed_at),
             datetime('now')
           );
         END;

         CREATE TRIGGER IF NOT EXISTS sync_trig_update_cards
         AFTER UPDATE ON cards
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_changes (id, device_id, table_name, row_id, op_type, logical_clock, payload_json, created_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'cards', NEW.id, 'update',
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             json_object('id',NEW.id,'language',NEW.language,'headword',NEW.headword,'answer_text',NEW.answer_text,'example_text',NEW.example_text,'notes',NEW.notes,'audio_reference',NEW.audio_reference,'source_entry_ids',NEW.source_entry_ids,'learning_status',NEW.learning_status,'created_at',NEW.created_at,'updated_at',NEW.updated_at,'last_reviewed_at',NEW.last_reviewed_at),
             datetime('now')
           );
         END;

         CREATE TRIGGER IF NOT EXISTS sync_trig_delete_cards
         AFTER DELETE ON cards
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_tombstones (id, device_id, table_name, row_id, logical_clock, deleted_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'cards', OLD.id,
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             datetime('now')
           );
         END;",
    )?;

    conn.execute_batch(
        "CREATE TRIGGER IF NOT EXISTS sync_trig_insert_collections
         AFTER INSERT ON collections
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_changes (id, device_id, table_name, row_id, op_type, logical_clock, payload_json, created_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'collections', NEW.id, 'insert',
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             json_object('id',NEW.id,'name',NEW.name,'description',NEW.description,'created_at',NEW.created_at,'updated_at',NEW.updated_at),
             datetime('now')
           );
         END;

         CREATE TRIGGER IF NOT EXISTS sync_trig_update_collections
         AFTER UPDATE ON collections
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_changes (id, device_id, table_name, row_id, op_type, logical_clock, payload_json, created_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'collections', NEW.id, 'update',
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             json_object('id',NEW.id,'name',NEW.name,'description',NEW.description,'created_at',NEW.created_at,'updated_at',NEW.updated_at),
             datetime('now')
           );
         END;

         CREATE TRIGGER IF NOT EXISTS sync_trig_delete_collections
         AFTER DELETE ON collections
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_tombstones (id, device_id, table_name, row_id, logical_clock, deleted_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'collections', OLD.id,
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             datetime('now')
           );
         END;",
    )?;

    conn.execute_batch(
        "CREATE TRIGGER IF NOT EXISTS sync_trig_insert_collection_memberships
         AFTER INSERT ON collection_memberships
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_changes (id, device_id, table_name, row_id, op_type, logical_clock, payload_json, created_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'collection_memberships', NEW.collection_id || ':' || NEW.card_id, 'insert',
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             json_object('collection_id',NEW.collection_id,'card_id',NEW.card_id,'created_at',NEW.created_at),
             datetime('now')
           );
         END;

         CREATE TRIGGER IF NOT EXISTS sync_trig_delete_collection_memberships
         AFTER DELETE ON collection_memberships
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_tombstones (id, device_id, table_name, row_id, logical_clock, deleted_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'collection_memberships', OLD.collection_id || ':' || OLD.card_id,
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             datetime('now')
           );
         END;",
    )?;

    conn.execute_batch(
        "CREATE TRIGGER IF NOT EXISTS sync_trig_insert_review_events
         AFTER INSERT ON review_events
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_changes (id, device_id, table_name, row_id, op_type, logical_clock, payload_json, created_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'review_events', NEW.id, 'insert',
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             json_object('id',NEW.id,'card_id',NEW.card_id,'result',NEW.result,'reviewed_at',NEW.reviewed_at),
             datetime('now')
           );
         END;",
    )?;

    conn.execute_batch(
        "CREATE TRIGGER IF NOT EXISTS sync_trig_insert_chat_conversations
         AFTER INSERT ON chat_conversations
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_changes (id, device_id, table_name, row_id, op_type, logical_clock, payload_json, created_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'chat_conversations', NEW.id, 'insert',
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             json_object('id',NEW.id,'title',NEW.title,'model',NEW.model,'collection_id',NEW.collection_id,'created_at',NEW.created_at,'updated_at',NEW.updated_at),
             datetime('now')
           );
         END;

         CREATE TRIGGER IF NOT EXISTS sync_trig_update_chat_conversations
         AFTER UPDATE ON chat_conversations
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_changes (id, device_id, table_name, row_id, op_type, logical_clock, payload_json, created_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'chat_conversations', NEW.id, 'update',
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             json_object('id',NEW.id,'title',NEW.title,'model',NEW.model,'collection_id',NEW.collection_id,'created_at',NEW.created_at,'updated_at',NEW.updated_at),
             datetime('now')
           );
         END;

         CREATE TRIGGER IF NOT EXISTS sync_trig_delete_chat_conversations
         AFTER DELETE ON chat_conversations
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_tombstones (id, device_id, table_name, row_id, logical_clock, deleted_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'chat_conversations', OLD.id,
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             datetime('now')
           );
         END;",
    )?;

    conn.execute_batch(
        "CREATE TRIGGER IF NOT EXISTS sync_trig_insert_chat_messages
         AFTER INSERT ON chat_messages
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_changes (id, device_id, table_name, row_id, op_type, logical_clock, payload_json, created_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'chat_messages', NEW.id, 'insert',
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             json_object('id',NEW.id,'conversation_id',NEW.conversation_id,'role',NEW.role,'body',NEW.body,'metadata_json',NEW.metadata_json,'created_at',NEW.created_at),
             datetime('now')
           );
         END;

         CREATE TRIGGER IF NOT EXISTS sync_trig_delete_chat_messages
         AFTER DELETE ON chat_messages
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_tombstones (id, device_id, table_name, row_id, logical_clock, deleted_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'chat_messages', OLD.id,
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             datetime('now')
           );
         END;",
    )?;

    conn.execute_batch(
        "CREATE TRIGGER IF NOT EXISTS sync_trig_insert_custom_chat_models
         AFTER INSERT ON custom_chat_models
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_changes (id, device_id, table_name, row_id, op_type, logical_clock, payload_json, created_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'custom_chat_models', NEW.id, 'insert',
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             json_object('id',NEW.id,'name',NEW.name,'title',NEW.title,'provider',NEW.provider,'created_at',NEW.created_at),
             datetime('now')
           );
         END;

         CREATE TRIGGER IF NOT EXISTS sync_trig_update_custom_chat_models
         AFTER UPDATE ON custom_chat_models
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_changes (id, device_id, table_name, row_id, op_type, logical_clock, payload_json, created_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'custom_chat_models', NEW.id, 'update',
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             json_object('id',NEW.id,'name',NEW.name,'title',NEW.title,'provider',NEW.provider,'created_at',NEW.created_at),
             datetime('now')
           );
         END;

         CREATE TRIGGER IF NOT EXISTS sync_trig_delete_custom_chat_models
         AFTER DELETE ON custom_chat_models
         WHEN (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1) IS NOT NULL
         BEGIN
           INSERT INTO sync_tombstones (id, device_id, table_name, row_id, logical_clock, deleted_at)
           VALUES (
             lower(hex(randomblob(16))),
             (SELECT id FROM sync_devices WHERE is_local = 1 LIMIT 1),
             'custom_chat_models', OLD.id,
             COALESCE((SELECT MAX(lc) FROM (SELECT MAX(logical_clock) AS lc FROM sync_changes UNION ALL SELECT MAX(logical_clock) AS lc FROM sync_tombstones)), 0) + 1,
             datetime('now')
           );
         END;",
    )?;

    Ok(())
}

File: apps/mobile/src/db/schema.ts
import {
  sqliteTable,
  text,
  integer,
  index,
  uniqueIndex,
  primaryKey,
} from 'drizzle-orm/sqlite-core';
import { sql } from 'drizzle-orm';

export const appMeta = sqliteTable('app_meta', {
  key:   text('key').primaryKey().notNull(),
  value: text('value'),
});

export const languages = sqliteTable('languages', {
  code:      text('code').primaryKey().notNull(),
  title:     text('title').notNull(),
  createdAt: text('created_at').notNull(),
});

export const dictionaries = sqliteTable('dictionaries', {
  id:             text('id').primaryKey().notNull(),
  name:           text('name').notNull().unique(),
  languageFrom:   text('language_from').notNull(),
  languageTo:     text('language_to').notNull(),
  sourceFilename: text('source_filename').notNull(),
  sourcePath:     text('source_path'),
  importStatus:   text('import_status', {
    enum: ['queued', 'importing', 'ready', 'failed'],
  }).notNull().default('queued'),
  entryCount:     integer('entry_count').notNull().default(0),
  lastError:      text('last_error'),
  createdAt:      text('created_at').notNull(),
  updatedAt:      text('updated_at').notNull(),
});

export const dictionaryEntries = sqliteTable('dictionary_entries', {
  id:                 text('id').primaryKey().notNull(),
  dictionaryId:       text('dictionary_id').notNull()
                        .references(() => dictionaries.id, { onDelete: 'cascade' }),
  headword:           text('headword').notNull(),
  normalizedHeadword: text('normalized_headword').notNull(),
  transcription:      text('transcription'),
  definitionText:     text('definition_text').notNull(),
  exampleText:        text('example_text'),
  audioReference:     text('audio_reference'),
  sourceOrder:        integer('source_order').notNull().default(0),
  createdAt:          text('created_at').notNull(),
}, (table) => [
  index('idx_dictionary_entries_dictionary_id').on(table.dictionaryId),
  index('idx_dictionary_entries_normalized_headword').on(table.normalizedHeadword),
]);

export const cards = sqliteTable('cards', {
  id:             text('id').primaryKey().notNull(),
  language:       text('language').notNull(),
  headword:       text('headword').notNull(),
  answerText:     text('answer_text').notNull(),
  exampleText:    text('example_text'),
  notes:          text('notes'),
  audioReference: text('audio_reference'),
  sourceEntryIds: text('source_entry_ids').notNull().default('[]'),
  learningStatus: text('learning_status', {
    enum: ['unreviewed', 'learned', 'not_learned'],
  }).notNull().default('unreviewed'),
  createdAt:      text('created_at').notNull(),
  updatedAt:      text('updated_at').notNull(),
  lastReviewedAt: text('last_reviewed_at'),
}, (table) => [
  uniqueIndex('ux_cards_headword_language').on(table.headword, table.language),
]);

export const collections = sqliteTable('collections', {
  id:          text('id').primaryKey().notNull(),
  name:        text('name').notNull().unique(),
  description: text('description'),
  createdAt:   text('created_at').notNull(),
  updatedAt:   text('updated_at').notNull(),
});

export const collectionMemberships = sqliteTable('collection_memberships', {
  collectionId: text('collection_id').notNull()
                  .references(() => collections.id, { onDelete: 'cascade' }),
  cardId:       text('card_id').notNull()
                  .references(() => cards.id, { onDelete: 'cascade' }),
  createdAt:    text('created_at').notNull(),
}, (table) => [
  primaryKey({ columns: [table.collectionId, table.cardId] }),
]);

export const reviewEvents = sqliteTable('review_events', {
  id:         text('id').primaryKey().notNull(),
  cardId:     text('card_id').notNull()
                .references(() => cards.id, { onDelete: 'cascade' }),
  result:     text('result', { enum: ['learned', 'not_learned'] }).notNull(),
  reviewedAt: text('reviewed_at').notNull(),
});

export const aiCredentials = sqliteTable('ai_credentials', {
  id:        text('id').primaryKey(),
  provider:  text('provider').notNull(),
  label:     text('label'),
  isActive:  integer('is_active', { mode: 'boolean' }).notNull().default(true),
  createdAt: text('created_at').notNull(),
  updatedAt: text('updated_at').notNull(),
}, (table) => [
  uniqueIndex('ux_ai_credentials_provider_active')
    .on(table.provider)
    .where(sql`${table.isActive} = 1`),
]);

export const chatConversations = sqliteTable('chat_conversations', {
  id:           text('id').primaryKey().notNull(),
  title:        text('title').notNull(),
  model:        text('model'),
  collectionId: text('collection_id'),
  createdAt:    text('created_at').notNull(),
  updatedAt:    text('updated_at').notNull(),
});

export const chatMessages = sqliteTable('chat_messages', {
  id:             text('id').primaryKey().notNull(),
  conversationId: text('conversation_id').notNull()
                    .references(() => chatConversations.id, { onDelete: 'cascade' }),
  role:           text('role', { enum: ['user', 'assistant'] }).notNull(),
  body:           text('body').notNull(),
  metadataJson:   text('metadata_json'),
  createdAt:      text('created_at').notNull(),
}, (table) => [
  index('idx_chat_messages_conversation_created')
    .on(table.conversationId, table.createdAt),
]);

export const customChatModels = sqliteTable('custom_chat_models', {
  id:        text('id').primaryKey(),
  name:      text('name').notNull().unique(),
  title:     text('title').notNull(),
  provider:  text('provider').notNull(),
  createdAt: text('created_at').notNull(),
});

// ---------------------------------------------------------------------------
// Sync metadata tables
// ---------------------------------------------------------------------------

/** Known local and remote device identities. Non-secret peer metadata only.
 *  Long-lived credentials are stored in expo-secure-store keyed by device id. */
export const syncDevices = sqliteTable('sync_devices', {
  id:          text('id').primaryKey().notNull(), // UUID device identifier
  displayName: text('display_name').notNull(),
  host:        text('host'),                      // null for the local device
  port:        integer('port'),                   // null for the local device
  isLocal:     integer('is_local', { mode: 'boolean' }).notNull().default(false),
  pairedAt:    text('paired_at'),
  lastSyncAt:  text('last_sync_at'),
  createdAt:   text('created_at').notNull(),
});

/** Local mutation log for rows that participate in sync v1.
 *  Covers INSERT and UPDATE operations. DELETEs go to sync_tombstones.
 *  Populated by SQLite triggers on synced domain tables. */
export const syncChanges = sqliteTable('sync_changes', {
  id:           text('id').primaryKey().notNull(),
  deviceId:     text('device_id').notNull(),
  tableName:    text('table_name').notNull(),
  rowId:        text('row_id').notNull(),
  opType:       text('op_type', { enum: ['insert', 'update'] }).notNull(),
  logicalClock: integer('logical_clock').notNull(),
  payloadJson:  text('payload_json').notNull(),
  requestId:    text('request_id'),    // set when received from a peer
  appliedAt:    text('applied_at'),    // null = unacknowledged local change
  createdAt:    text('created_at').notNull(),
}, (table) => [
  index('idx_sync_changes_device_clock').on(table.deviceId, table.logicalClock),
  index('idx_sync_changes_table_row').on(table.tableName, table.rowId),
  index('idx_sync_changes_applied_at').on(table.appliedAt),
]);

/** Delete records for rows that must remain deleted across peers. */
export const syncTombstones = sqliteTable('sync_tombstones', {
  id:           text('id').primaryKey().notNull(),
  deviceId:     text('device_id').notNull(),
  tableName:    text('table_name').notNull(),
  rowId:        text('row_id').notNull(),
  logicalClock: integer('logical_clock').notNull(),
  requestId:    text('request_id'),
  deletedAt:    text('deleted_at').notNull(),
}, (table) => [
  index('idx_sync_tombstones_device_clock').on(table.deviceId, table.logicalClock),
  index('idx_sync_tombstones_table_row').on(table.tableName, table.rowId),
]);

/** Last acknowledged cursor per peer device, per direction. */
export const syncCursors = sqliteTable('sync_cursors', {
  peerDeviceId:                   text('peer_device_id').primaryKey().notNull(),
  lastRemoteLogicalClock:         integer('last_remote_logical_clock').notNull().default(0),
  lastAcknowledgedLocalLogicalClock: integer('last_acknowledged_local_logical_clock').notNull().default(0),
  updatedAt:                      text('updated_at').notNull(),
});

/** Singleton row tracking current pairing state and last sync result. */
export const syncState = sqliteTable('sync_state', {
  id:               text('id').primaryKey().notNull().default('local'),
  pairedDesktopId:  text('paired_desktop_id'), // references syncDevices.id
  lastSyncAt:       text('last_sync_at'),
  lastSyncResult:   text('last_sync_result', { enum: ['success', 'failure'] }),
  lastSyncError:    text('last_sync_error'),
  lastRequestId:    text('last_request_id'),
  /** Set once after the first successful sync for this trust period; cleared on re-pair / forget / revoke. */
  firstSuccessfulSyncAt: text('first_successful_sync_at'),
  updatedAt:        text('updated_at').notNull(),
});
